### PR TITLE
Deploy serving image into Cloud Run Service

### LINKED ISSUE #18: Update github actions to match workflow
Yes, that needs changes, reflecting our views on the process described by Michel.

_Originally posted by @anayden in https://github.com/mlops-labs-team1/engineering.labs/issues/7#issuecomment-768974545_

### PR DESCRIPTION
**Pull request by michelav**
Deploy Torchserve image with embedded model into Cloud Run Service. The branch contains the updates from _train-build_.

### CODE REVIEW

**Comment by michelav** [comment]
I'm using the infra created with TF. I've followed the steps from _train-build_ branch and I could push the image to Google Container Registry. I'm getting the following error when deploying to Cloud Run: `Cloud Run error: Container failed to start. Failed to start and then listen on the port defined by the PORT environment variable.` Still investigating but maybe Cloud Run couldn't confirm that container is listening the the configured port: _8080_. I'll try to set up torchserve instance using `--ts-config <path_to_config>/config.properties`.

**Comment by michelav** [comment]
**Here the train-build-push-deploy process (I'm using the TF Infra):**

1. Cloned the repo
  ```
  git clone --single-branch cloud-run-serving https://github.com/mlops-labs-team1/engineering.labs.git
 ```

2. Removed GCP and secrets part from Dockerfile-trainer

3. Renamed image on MLproject file (only for tests)

4. Build training image
```
 docker build -t englabs/bert-trainer -f ./Dockerfile-trainer .`
```

5. Generate json file that will hold information from model
 ```
JSON_FILE="$(uuidgen).json"
```

6. Run the training
```
mlflow run . -P json_dump=/tmp/englab/$JSON_FILE -P model_name=BertModel -P max_epochs=1 -P num_samples=64
```

7 . Extracting data from json file
   ```
   NAME=$(cat ${JSON_FILE} |jq -r '.name')
   VERSION=$(cat ${JSON_FILE} |jq -r '.version')
   MODEL_NAME=/${NAME}/$VERSION
   MODEL_TAG=${NAME}-$VERSION
   ```

8. Removed from Dokerfile-torchserve: GCP_CREDS_JSON_BASE64_ENV and GOOGLE_APPLICATION_CREDENTIALS. (Credentials come from service account registered in VM)

8. Created an ARG to set MLFLOW_TRACKING_URI in _Dockerfile-torchserve_

9. Build serving image
```
docker build -t mlflow-torch-test:${MODEL_TAG} --build-arg MODEL_NAME=${MODEL_NAME} --build-arg MLFLOW_TRACKING_URI_ARG=${MLFLOW_TRACKING_URI} -f ./Dockerfile-torchserve .
```

10. Tag the image before pushing
```
docker tag mlflow-torch-test:${MODEL_TAG} gcr.io/engineeringlab/mlflow-torch-test:${MODEL_TAG}
```

11. Added role `"roles/run.admin"`, `"roles/iam.serviceAccountUser"` and `"roles/storage.admin"` to service account (Already done via TF> Will be ok next run)

12. Push serve image to Registry
```
docker push gcr.io/engineeringlab/mlflow-torch-test:${MODEL_TAG}
```

13. Trying to deploy to Cloud Run (Set all those parameters to avoid requests from the command through STDIN) 
```
gcloud run deploy ${NAME,,}-service --image gcr.io/engineeringlab/mlflow-torch-test:${MODEL_TAG} --platform managed --port 8080 --allow-unauthenticated --memory 512M --region southamerica-east1
```

After that, got this error:
**Cloud Run error: Container failed to start. Failed to start and then listen on the port defined by the PORT environment variable.**

**Comment by michelav** [comment]
It works! Command used to deploy container in Cloud Run:

`gcloud run deploy news-bert-service --image gcr.io/engineeringlab/labs1-mlflow-torch:BertModel-26 --platform managed --port 8080 --allow-unauthenticated --memory 2G --cpu 2 --region southamerica-east1`

File: Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/Dockerfile-torchserve [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.8.1-slim
+FROM python:3.7
```
Comment (savagej) [inline Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/Dockerfile-torchserve:1]: An idea for another PR (or maybe for #20 ), since we're so dependent on pickle for ser/des, would it make sense to have the same base image for both the trainer and the server so that there's much less chance of differences between them?
Another advantage of this is that the `pytorch/serve` repo's `install_dependencies.py` is pretty smart and will install the correct version of pytorch based on the system and cuda version available: https://github.com/pytorch/serve/blob/master/ts_scripts/install_dependencies.py#L25-L36, which means our trainer would have a more optimised version of pytorch based on where it spins up.

File: Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/MLproject [hunk 2]
```diff
@@ -2,7 +2,8 @@ name: bert-classification
 
 docker_env:
   image: englabs/trainer
-  environment: [["MLFLOW_TRACKING_URI", "http://35.185.111.8:5000"]]
+  volumes: ["$(pwd):/tmp/englab/"]
+  environment: ["MLFLOW_TRACKING_URI"]
```
Comment (savagej) [inline Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/MLproject:6]: I don't know if we need this any more, I think `mlflow run` is smart enough to use the host version itself if defined or use the one in the docker image if not

**Review by savagej — APPROVED** [review]

File: Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/Dockerfile-torchserve (hunk 1 above)
Comment (savagej) [reply Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/Dockerfile-torchserve:1]: By base image I mean we could do something like create a Dockerfile.base for installing python dependencies, then have two other files for the trainer and server that start FROM that base. Could be overkill but we've already had a lot of heartache with differences between train/serve

Comment (michelav) [reply Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/Dockerfile-torchserve:1]: > An idea for another PR (or maybe for #20 ), since we're so dependent on pickle for ser/des, would it make sense to have the same base image for both the trainer and the server so that there's much less chance of differences between them?

Definitely makes sense!

File: Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/MLproject (hunk 2 above)
Comment (michelav) [reply Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/MLproject:6]: Yes it is. However you gotta tell him to copy the variable if it exists in the host. That's why I let the _env_ there without its value.
